PennMobile/Courses/Models/Course.swift

```diff
@@ -0,0 +1,150 @@
+
+    /// An array of meeting times for the course.
+    var meetingTimes: [MeetingTime]?
+}
+
+// Literally here so SwiftUI is happy
+extension Course: Identifiable {
+    var id: String { crn }
+}
+
+extension Course {
+    enum ConversionError: Error {
+        case invalidWeekday
+        case invalidTimeString
+    }
+
+    /// Initializes a course from Path@Penn data.
+    init(_ data: PathAtPennNetworkManager.CourseData) {
+        crn = data.crn
+        code = data.code
+        title = data.title
+        section = data.section
+
+        do {
+            let instructorHTML = try SwiftSoup.parse(data.instructordetail_html)
+            let divs = try instructorHTML.select("div")
+            instructors = try divs.map { try $0.text(trimAndNormaliseWhitespace: true) }
+        } catch let error {
+            instructors = []
+            print("Couldn't parse instructors: \(error)")
+        }
+
+        do {
+            let meetingHTML = try SwiftSoup.parse(data.meeting_html)
+            let a = try meetingHTML.select("a").first()
+            location = try a?.text(trimAndNormaliseWhitespace: true)
+        } catch let error {
+            location = nil
+            print("Couldn't parse meeting HTML: \(error)")
+        }
+
+        struct PathAtPennMeetingTime: Decodable {
+            var meet_day: String
+            var start_time: String
+            var end_time: String
+        }
+
+        if let sectionData = data.allInGroup.first(where: { $0.crn == data.crn }) {
+            startDate = sectionData.start_date
+            endDate = sectionData.end_date
+
+            do {
+                let timeData = try sectionData.meetingTimes.data(using: .utf8).unwrap(orThrow: PathAtPennError.corruptString)
+                let rawTimes = try JSONDecoder().decode([PathAtPennMeetingTime].self, from: timeData)
+                meetingTimes = rawTimes.compactMap { time in
+                    // Path@Penn returns 0 through 6 for Monday to Sunday
+                    // We need to map that to 1 through 7 for Sunday to Saturday
+                    guard let dayInt = Int(time.meet_day), (0...6).contains(dayInt) else {
+                        print("Got invalid weekday: \(time.meet_day)")
+                        return nil
+                    }
+                    let weekday = (dayInt + 2) % 7
+
+                    do {
+                        let start = try getTimeInt(pathAtPennString: time.start_time)
+                        let end = try getTimeInt(pathAtPennString: time.end_time)
+                        return MeetingTime(weekday: weekday, startTime: start, endTime: end)
+                    } catch let error {
+                        print("Couldn't parse start/end times: \(error)")
+                        return nil
+                    }
+                }
+            } catch let error {
+                meetingTimes = []
+                print("Couldn't parse meeting times: \(error)")
+            }
```

Comment:
is the nested `do` `catch` block necessary here as well?

Reply:
Yeah, it's probably cleaner without it.